scripts/notebooks/Batching.ipynb

```diff
@@ -0,0 +1,1275 @@
+    "    while True:\n",
+    "        batches_to_adjust = [b for b in batches if len(batches[b]) < min_batch_size or len(batches[b]) > max_batch_size]\n",
+    "        if not batches_to_adjust:\n",
+    "            break\n",
+    "        \n",
+    "        sorted_batches = sorted(batches.items(), key=lambda x: len(x[1]))\n",
+    "        for batch_name in batches_to_adjust:\n",
+    "            if len(batches[batch_name]) < min_batch_size:\n",
+    "                samples_needed = min_batch_size - len(batches[batch_name])\n",
+    "                for large_batch_name, large_batch_samples in reversed(sorted_batches):\n",
+    "                    if len(large_batch_samples) > min_batch_size:\n",
+    "                        samples_to_move = min(samples_needed, len(large_batch_samples) - min_batch_size)\n",
+    "                        batches[batch_name].extend(large_batch_samples[-samples_to_move:])\n",
+    "                        batches[large_batch_name] = large_batch_samples[:-samples_to_move]\n",
+    "                        samples_needed -= samples_to_move\n",
+    "                        if samples_needed == 0:\n",
+    "                            break\n",
+    "            \n",
+    "            elif len(batches[batch_name]) > max_batch_size:\n",
+    "                overflow = batches[batch_name][max_batch_size:]\n",
+    "                batches[batch_name] = batches[batch_name][:max_batch_size]\n",
+    "                for small_batch_name, small_batch_samples in sorted_batches:\n",
+    "                    if len(small_batch_samples) < max_batch_size:\n",
+    "                        space_available = max_batch_size - len(small_batch_samples)\n",
+    "                        samples_to_move = min(space_available, len(overflow))\n",
+    "                        batches[small_batch_name].extend(overflow[:samples_to_move])\n",
+    "                        overflow = overflow[samples_to_move:]\n",
+    "                        if not overflow:\n",
+    "                            break\n",
+    "                if overflow:\n",
+    "                    batch_num += 1\n",
+    "                    new_batch_name = f\"{batch_prefix}{batch_num}{batch_suffix}\"\n",
+    "                    batches[new_batch_name] = overflow\n",
+    "        sorted_batches = sorted(batches.items(), key=lambda x: len(x[1]))\n",
+    "    \n",
+    "    return batches, batches_meta"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Data Ingestion\n",
+    "This section fetches the sample data."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "code_folding": []
+   },
+   "outputs": [],
+   "source": [
+    "# Display metadata table, which was generated as part of EvidenceQC\n",
+    "PASS_METADATA = os.path.join(WS_BUCKET, \"evidence_qc/filtering/passed_samples_qc_metadata.tsv\")\n",
+    "pass_df = pd.read_table(PASS_METADATA)\n",
```

Comment:
This is breaking on me:
```
FileNotFoundError: b/fc-41a63557-9ddf-41ba-b5d0-807be90ddb87/o/evidence_qc%2Ffiltering%2Fpassed_samples_qc_metadata.tsv
```
Does pandas support reading directly from buckets?

Reply:
Yes, it does - you need to use `gs://` as a prefix though.

Reply:
Sorry, on closer inspection I realized that this is automatically set based on `WS_BUCKET`, which is pulled from `os.environ['WORKSPACE_BUCKET']`. Strange to me that this does not automatically prepend `gs://` in this case though, as it does for me...

Reply:
Updated accordingly - file path now matches what is generated in the QC notebook.